### PR TITLE
fix(dynamite): too long names for someOf typedefs

### PR DESCRIPTION
**Pull request by Leptopoda**

### CODE REVIEW

**Review by provokateurin — COMMENTED** [review]
Ugh ok no we can't do this, it's a horrible user experience. We are just working around a problem with dart doc here, we need to fix it there.

**Comment by Leptopoda** [comment]
What do you mean?

In theory we can make them visibleForTesting. Users would loose the ability to have a `fromJson` reviver but I doubt that  this is a problem.
I also don't think that there is a way to fix this other than dartdoc generating html files with hashed names and using some sort of javaScript routing to handle the correct url client side. And I don't see them investing is such work.

File: packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart [hunk 1]
```diff
@@ -58,30 +58,31 @@ abstract class NestedRedirect implements $NestedRedirectInterface, Built<NestedR
 }
 
 typedef SomeOfRedirect = ({Base? base, int? $int, JsonObject? jsonObject});
-typedef $BaseIntJsonObject = ({Base? base, int? $int, JsonObject? jsonObject});
+typedef $e1c7ecea8e5fdae7b94cd86c0dc4f1ba = ({Base? base, int? $int, JsonObject? jsonObject});
```
Comment (Leptopoda) [inline packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: As one can see here there is always going to be a typedef with a usable name.
The hashed name is only needed for the serializer to match all typedefs with those usable names.

**Comment by provokateurin** [comment]
Well in some places one needs to create objects, e.g. spreed signaling. Using some weird hash name is not ok there

**Comment by Leptopoda** [comment]
> Well in some places one needs to create objects, e.g. spreed signaling. Using some weird hash name is not ok there

Please see the above code comment.

File: packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart (hunk 1 above)
Comment (provokateurin) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: I disagree about the "usable". Also in the generate docs, the names would be meaningless. There's no way we can push this onto the package consumers.

Comment (provokateurin) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: I'd say not having a typedef is more user friendly in this case.

Comment (Leptopoda) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: The problem is that we need the typedef for the serializer as it can not work with raw records (the runtimetype of them is not constant while the one of a typedef is. So it also has performance implications).

Therefore we generate a typedef for use by the devs (`SomeOfRedirect` and one to use by the serializer `$e1c7ecea8e5fdae7b94cd86c0dc4f1ba`).
A dev should never need to use the later one as everything (other than the fromJson method that's static to the extension).

Comment (Leptopoda) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: If the fromJson reviver is that important We can generate extensions for the human readable version to only handle that while having a common extension (on the hashed name) that handles the non static methods.

Comment (Leptopoda) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: And if you are curious we can not generate extension methods with the non static methods for every typedef as ones this would lead to conflicts between someOfs with the same type.

Comment (provokateurin) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: Ok, but how would you generate human readable extension names that are not conflicting and not too long?

Comment (Leptopoda) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart:61]: We already do generate human readable ones.
That's the whole point.

The not human readable ones are not needed by the devs. I just made them private and will push these changes later. Maybe this helps in understanding.

Comment (Leptopoda) [reply packages/dynamite/dynamite_end_to_end_test/lib/type_defs.openapi.dart]: Sorry, I didn't find the time earlier.
Does it make sense now?

**Comment by provokateurin** [comment]
Just to be sure I understand the approach: The serializers and records are private and hashed, but we have proxy extensions that are human readable. There might be multiple proxy extensions pointing to the same type as they would otherwise have to have the long names we had problems. Is this correct?

One problem I still see is that someone could create a schema that has >255 characters, but we can't do much about that and it would be quite stupid anyway.

**Comment by Leptopoda** [comment]
> One problem I still see is that someone could create a schema that has >255 characters, but we can't do much about that and it would be quite stupid anyway.

The good thing is that dynamite would not fail. They would just not be able to create documentation for their generated code (which is out of our responsibility).

**Comment by provokateurin** [comment]
Can you confirm that I understand this change correctly?

**Comment by provokateurin** [comment]
The commit message has a typo btw (`too` not `to`)

**Comment by Leptopoda** [comment]
Ah sorry.
Yes you understand that correctly.

File: cspell.json [hunk 2]
```diff
@@ -11,7 +11,9 @@
     "**.mocks.dart",
     "**/assets",
     "**/l10n/!(en.arb)",
+    "**/*.openapi.dart",
```
Comment (provokateurin) [inline cspell.json:14]:
```suggestion
    "**.openapi.dart",
```

File: cspell.json [hunk 3]
```diff
@@ -11,7 +11,9 @@
...
     "external",
+    "packages/dynamite/dynamite_end_to_end_test/test/*.dart",
```
Comment (provokateurin) [inline cspell.json:16]: This shouldn't be needed anymore

**Comment by Leptopoda** [comment]
> The commit message has a typo btw (`too` not `to`)

Is the rest fine? So I only need to push once more?

File: packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart [hunk 4]
```diff
@@ -17,7 +17,7 @@ String toDartName(
     }
   }
 
-  if (_reservedNames.contains(result) || RegExp(r'^[0-9]+$', multiLine: true).hasMatch(result)) {
+  if (_reservedNames.contains(result) || RegExp(r'^[0-9]+$', multiLine: true).hasMatch(result[0])) {
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart:20]: Why this change?

File: packages/dynamite/dynamite/pubspec.yaml [hunk 5]
```diff
@@ -11,6 +11,7 @@ dependencies:
   checked_yaml: ^2.0.0
   code_builder: ^4.5.0
   collection: ^1.0.0
+  crypto: ^3.0.3
```
Comment (provokateurin) [inline packages/dynamite/dynamite/pubspec.yaml:14]:
```suggestion
  crypto: ^3.0.0
```
Should be enough

File: packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart [hunk 6]
```diff
@@ -112,8 +137,9 @@ Iterable<Spec> generateSomeOf(
 
   yield Extension(
     (final b) => b
-      ..name = '${identifier}Extension'
+      ..name = '${identifier}Extension'.nonPrivate
       ..on = refer(identifier)
+      ..docs.add('// ignore: library_private_types_in_public_api')
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: The whole extension could be private so we don't need this, right?

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: Then we also wouldn't need to hash the name as it won't appear in the docs anyway :thinking:

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart (hunk 4 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart:20]: Not needed anymore as we now prefix all hashed names with a `$`.
I'll put it in a separate PR

File: packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart (hunk 6 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: This would also hide the `toJson` and `validate` methods` which might be interesting to devs.

File: packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart (hunk 4 above)
Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/helpers/dart_helpers.dart:20]: But then the expression will always be false because it never starts with a number?

File: packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart (hunk 6 above)
Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: But we could expose those through the proxy extension? I like this way more as it will solve this problem the right way imo

Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: But we could expose those through the proxy extension? 
Nope as you would have multiple extension declaring the same method.

Which one should be used?
```dart
print('value'.name);

extension Extension1 on String{
  String get name => '1';
}

extension Extension2 on String{
  String get name => '2';
}
```
I know that one could explicitly specify this like `Extension1('value').name`. But that's tedious.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: Ah I see :/
I would still prefer to make the whole extension private. We expose the serializers so if someone really needs toJson they can do it manually.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: And we can still make it public later, it wouldn't be a breaking change. Keeping it public now and making it private later will be a breaking change.

Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: The seriailizers with all registered builder factories needed are still annotated with `visibleForTesting`.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: But that we can also change later without it being a breaking change if necessary.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: But I realized in this case we can also argue against the proxy extension and just make it all private. If serializing/deserializing is necessary then we need to expose the jsonSerializers.

Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: I think the current implementation is fine.
We can still add doc comments to the ugly extension later.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/generate_ofs.dart:142]: But do we really want to have the hashed extensions appear in the docs? Is there maybe an annotation that removes it from the docs but makes it still accessible (not like `@internal`)?

**Review by provokateurin — APPROVED** [review]
Please fix the typo in the commit message and pr title before merging

**Comment by Leptopoda** [comment]
Can I get a new review?
I made the serializer getter private and added one per human readable name.
As it was static this would have been one (rare) case, where a dev might have encountered the hashed name.

**Review by provokateurin — APPROVED** [review]
